### PR TITLE
added Archiver.get/setLastModifiedDate(Date) API to force entries value

### PR DESCRIPTION
**Pull request by hboutemy**
this API is a step to Reproducible Builds, to get reproducible archives

it is based on comments made for #49

### CODE REVIEW

File: src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java [hunk 1]
```diff
@@ -1133,4 +1139,10 @@ public void setIgnorePermissions( final boolean ignorePermissions )
         this.ignorePermissions = ignorePermissions;
     }
 
+    @Override
+    public void setLastModifiedDate( Date lastModifiedDate ) { this.lastModifiedDate = lastModifiedDate; }
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: This does not correspond to the given format style

File: src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java [hunk 2]
```diff
@@ -1133,4 +1139,10 @@ public void setIgnorePermissions( final boolean ignorePermissions )
...
+
+    @Override
+    public Date getLastModifiedDate() { return lastModifiedDate; }
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/AbstractArchiver.java]: This does not correspond to the given format style

File: src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java [hunk 3]
```diff
@@ -331,4 +332,10 @@ public void setIgnorePermissions( boolean ignorePermissions )
         target.setIgnorePermissions( ignorePermissions );
     }
 
+    @Override
+    public void setLastModifiedDate( final Date lastModifiedDate ) { target.setLastModifiedDate(lastModifiedDate); }
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java]: This does not correspond to the given format style

File: src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java [hunk 4]
```diff
@@ -331,4 +332,10 @@ public void setIgnorePermissions( boolean ignorePermissions )
...
+
+    @Override
+    public Date getLastModifiedDate() { return target.getLastModifiedDate(); }
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java]: This does not correspond to the given format style

File: src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java [hunk 5]
```diff
@@ -18,6 +18,7 @@
 import java.io.File;
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java]: Side note: the class name has a typo

File: src/main/java/org/codehaus/plexus/archiver/diags/NoOpArchiver.java [hunk 6]
```diff
@@ -346,4 +347,13 @@ public void setIgnorePermissions( boolean ignorePermissions )
...
     }
 
+    @Override
+    public void setLastModifiedDate( final Date lastModifiedDate )
+    {
+
+    }
+
+    @Override
+    public Date getLastModifiedDate() { return null; }
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/NoOpArchiver.java]: This does not correspond to the given format style

File: src/main/java/org/codehaus/plexus/archiver/diags/TrackingArchiver.java [hunk 7]
```diff
@@ -400,4 +403,10 @@ public void setIgnorePermissions( final boolean ignorePermissions )
         this.ignorePermissions = ignorePermissions;
     }
 
+    @Override
+    public void setLastModifiedDate( final Date lastModifiedDate ) { this.lastModified = lastModifiedDate; }
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/TrackingArchiver.java]: This does not correspond to the given format style

File: src/main/java/org/codehaus/plexus/archiver/diags/TrackingArchiver.java [hunk 8]
```diff
@@ -400,4 +403,10 @@ public void setIgnorePermissions( final boolean ignorePermissions )
...
+
+    @Override
+    public Date getLastModifiedDate() { return lastModified; }
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/TrackingArchiver.java]: This does not correspond to the given format style

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java [hunk 9]
```diff
@@ -52,7 +52,7 @@
 {
     "NullableProblems", "UnusedDeclaration"
 } )
-public abstract class AbstractZipArchiver
+public abstract class  AbstractZipArchiver
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Redundant space

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java [hunk 10]
```diff
@@ -530,6 +530,11 @@ public InputStream get()
 
     private void setTime( java.util.zip.ZipEntry zipEntry, long lastModified )
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Shouldn't a specific setter don't override a global value?

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java (hunk 9 above)
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: nice catch: I could not see it while reviewing before committing :)

File: src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java (hunk 10 above)
Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: it's up to the method writer to define the semantics
coding the global override in the helper just makes the change easy, and avoids missing any location
if you want, we can change the name of the variable (originalLastModified?): but I can't find a good one...

Comment (michael-o) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: Regardless of the name, this has to be clearly documented. Maybe an overloaded method with boolean could cover both cases?

Comment (hboutemy) [reply src/main/java/org/codehaus/plexus/archiver/zip/AbstractZipArchiver.java]: hey, it's a private helper that is quite short: don't do too much

File: src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java [hunk 11]
```diff
@@ -331,4 +332,16 @@ public void setIgnorePermissions( boolean ignorePermissions )
         target.setIgnorePermissions( ignorePermissions );
     }
 
+    @Override
+    public void setLastModifiedDate( final Date lastModifiedDate )
+    {
+        target.setLastModifiedDate(lastModifiedDate);
```
Comment (michael-o) [inline src/main/java/org/codehaus/plexus/archiver/diags/DelgatingArchiver.java]: Formatting

**Pull request merged**